### PR TITLE
feat: add mix tasks for generating release artifacts

### PR DESCRIPTION
**Pull request by abc3**
This PR adds new mix tasks that facilitate the generation of release artifacts.

[Distillery's implementation ](https://github.com/bitwalker/distillery)is used for generating the `appup` file and `relup` is generated via `:systools.make_relup/4`.

### CODE REVIEW

File: lib/tasks/gen.appup.ex [hunk 1]
```diff
@@ -0,0 +1,51 @@
+defmodule Mix.Tasks.Supavisor.Gen.Appup do
+  @moduledoc """
+  Generates an appup file for a given release.
+
+  ## Examples
+
+    # Generate an appup from 0.0.1 to 0.0.2 versions
+
+    mix supavisor.gen.appup --from=0.0.1 --to=0.0.2
```
Comment (wojtekmach) [inline lib/tasks/gen.appup.ex:9]: 4 space indent for code blocks:

```suggestion
      # Generate an appup from 0.0.1 to 0.0.2 versions

      mix supavisor.gen.appup --from=0.0.1 --to=0.0.2
```

File: lib/tasks/gen.tarball.ex [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+  ## Examples
+
+    # Generate a tarball for version 0.0.1
+
+    mix supavisor.gen.tarball --vsn=0.0.1
+  """
+
+  use Mix.Task
+
+  @spec run(OptionParser.argv()) :: no_return
```
Comment (wojtekmach) [inline lib/tasks/gen.tarball.ex:14]:
```suggestion
  @impl true
```

:)

File: lib/tasks/gen.tarball.ex [hunk 3]
```diff
@@ -0,0 +1,96 @@
+defmodule Mix.Tasks.Supavisor.Gen.Tarball do
```
Comment (wojtekmach) [inline lib/tasks/gen.tarball.ex:1]: I believe the same can be achieved by adding `:tar` to release `:steps` in mix.exs?

**Review by wojtekmach — COMMENTED** [review]
LGTM

File: lib/tasks/gen.tarball.ex (hunk 3 above)
Comment (abc3) [reply lib/tasks/gen.tarball.ex:1]: Oh, interesting, I'm gonna read mix manual 🤔

Comment (wojtekmach) [reply lib/tasks/gen.tarball.ex:1]: Yup!

```elixir
  def project do
    [
      # ...
      releases: [
        foo: [
          steps: [
            :assemble,
            &foo/1,
            :tar
          ]
        ]
      ]
    ]
  end

  defp foo(release) do
    Mix.Task.run("eval", ["IO.inspect 1 + 2"])
    release
  end
```

**Comment by josevalim** [comment]
Also, for completeness, we talk about the challenges for hot code upgrades here: https://hexdocs.pm/mix/Mix.Tasks.Release.html#module-hot-code-upgrades :)

**Comment by abc3** [comment]
@josevalim, thanks for pointing out this notice to me. We want to use the built-in upgrading for minor changes and when it doesn't involve state changes. Our goal is to reduce user disconnection during updates when it is possible

**Review by chasers — APPROVED** [review]

**Pull request merged**